packages/react-native/Libraries/Blob/URL.js

```diff
@@ -114,96 +114,104 @@ export class URLSearchParams {
-      if (baseUrl.endsWith(url)) {
-        url = '';
-      }
-      this._url = `${baseUrl}${url}`;
-    }
-  }
+function resolveRelativeUrl(relative: string, base: string): string {
+  const baseUrl = new URL(base);
 
-  get hash(): string {
-    throw new Error('URL.hash is not implemented');
+  if (relative.startsWith('http://') || relative.startsWith('https://')) {
+    return relative;
   }
 
-  get host(): string {
-    throw new Error('URL.host is not implemented');
+  if (relative.startsWith('/')) {
+    return baseUrl.protocol + '//' + baseUrl.host + encodeURI(relative);
   }
 
-  get hostname(): string {
-    throw new Error('URL.hostname is not implemented');
-  }
+  const baseParts = baseUrl.pathname.split('/');
+  const relativeParts = relative.split('/');
 
-  get href(): string {
-    return this.toString();
-  }
+  baseParts.pop();
 
-  get origin(): string {
-    throw new Error('URL.origin is not implemented');
+  for (const part of relativeParts) {
+    if (part === '.') continue;
+    if (part === '..') baseParts.pop();
+    else baseParts.push(part);
   }
 
-  get password(): string {
-    throw new Error('URL.password is not implemented');
-  }
+  return baseUrl.protocol + '//' + baseUrl.host + baseParts.join('/');
+}
 
-  get pathname(): string {
-    throw new Error('URL.pathname not implemented');
-  }
+export class URL {
+  protocol: string;
+  username: string;
+  password: string;
+  host: string;
+  hostname: string;
+  port: string;
+  pathname: string;
+  search: string;
+  hash: string;
+  origin: string;
+  _url: string;
+  _searchParamsInstance: ?URLSearchParams = null;
 
-  get port(): string {
-    throw new Error('URL.port is not implemented');
-  }
+  constructor(url: string, base?: string | URL) {
+    if (base) {
+      if (typeof base === 'string') {
+        base = new URL(base);
+      }
+      url = resolveRelativeUrl(url, base.href);
+    } else {
+      url = encodeURI(url);
+    }
 
-  get protocol(): string {
-    throw new Error('URL.protocol is not implemented');
-  }
+    const parser = this.parseURL(url);
+    this._url = url;
+    this.protocol = parser.protocol;
+    this.username = parser.username;
+    this.password = parser.password;
+    this.host = parser.host;
+    this.hostname = parser.hostname;
+    this.port = parser.port;
+    this.pathname = parser.pathname;
+    this.search = parser.search;
+    this.hash = parser.hash;
+    this.origin = parser.origin;
+
+    if (this.pathname === '/' && !this.href.endsWith('/')) {
+      this._url += '/';
+    }
 
-  get search(): string {
-    throw new Error('URL.search is not implemented');
+    this._searchParamsInstance = new URLSearchParams(this.search);
+  }
+
+  parseURL(url: string): {
+    protocol: string,
+    username: string,
+    password: string,
+    host: string,
+    hostname: string,
+    port: string,
+    pathname: string,
+    search: string,
+    hash: string,
+    origin: string,
+  } {
+    const urlPattern =
```

Comment:
Where does this regex come from?

At a quick glance, I know we should be more permissive in allowing schemes that are not just http.

Reply:
@NickGerleman Oh I almost missed that, I'm not good enough to see that, thanks for the advice.

Reply:
Are we getting this regex from a source, or trying to derive it?

Reply:
@NickGerleman No, I didn't get this regex from a source. I derived it myself.